File: library/src/main/java/me/shkschneider/skeleton/data/JsonParser.java
package me.shkschneider.skeleton.data;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import me.shkschneider.skeleton.helper.LogHelper;

public class JsonParser {

    protected JsonParser() {
        // Empty
    }

    @Nullable
    public static JSONObject parse(@NonNull final String string) {
        try {
            return new JSONObject(string);
        }
        catch (final JSONException e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    @Nullable
    public static JSONObject parse(@NonNull final InputStream inputStream) {
        final String string = FileHelper.readString(inputStream);
        if (TextUtils.isEmpty(string)) {
            LogHelper.warning("String was NULL");
            return null;
        }
        return parse(string);
    }

    @SuppressWarnings("unchecked")
    @Nullable
    public static <T> T get(@NonNull final JSONObject jsonObject, @NonNull final String key, final T fallback) {
        try {
            return (T) jsonObject.get(key);
        }
        catch (final JSONException e) {
            return fallback;
        }
    }

    @SuppressWarnings("unchecked")
    @Nullable
    public static <T> T get(@NonNull final JSONArray jsonArray, final int index, final T fallback) {
        try {
            return (T) jsonArray.get(index);
        }
        catch (final JSONException e) {
            return fallback;
        }
    }

    public static List<String> keys(@NonNull final JSONObject jsonObject) {
        final List<String> keys = new ArrayList<>();
        final Iterator<String> iterator = jsonObject.keys();
        while (iterator.hasNext()) {
            keys.add(iterator.next());
        }
        return keys;
    }

    public static List<Object> values(@NonNull final JSONObject jsonObject) {
        final List<Object> values = new ArrayList<>();
        try {
            for (final String key : keys(jsonObject)) {
                values.add(jsonObject.get(key));
            }
        }
        catch (final JSONException e) {
            LogHelper.wtf(e);
        }
        return values;
    }

    public static boolean has(@NonNull final JSONObject jsonObject, @NonNull final String key) {
        return jsonObject.has(key);
    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/data/JsonParser.kt
package me.shkschneider.skeleton.data

import me.shkschneider.skeleton.helper.Logger
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.*

@Deprecated("Use Gson library.")
object JsonParser : IParser<JSONObject, JSONArray> {

    override fun parse(string: String): JSONObject? {
        return try {
            JSONObject(string)
        } catch (e: JSONException) {
            Logger.wtf(e)
            null
        }
    }

    override fun keys(jsonObject: JSONObject): List<String> {
        val keys = ArrayList<String>()
        val iterator = jsonObject.keys()
        while (iterator.hasNext()) {
            keys.add(iterator.next())
        }
        return keys
    }

    override fun values(jsonObject: JSONObject): List<Any> {
        val values = ArrayList<Any>()
        try {
            keys(jsonObject).forEach { key ->
                values.add(jsonObject.get(key))
            }
        } catch (e: JSONException) {
            Logger.wtf(e)
        }
        return values
    }

    override fun has(jsonObject: JSONObject, key: String): Boolean {
        return jsonObject.has(key)
    }

    // from Object

    override fun get(jsonObject: JSONObject, key: String, fallback: JSONObject?): JSONObject? {
        if (jsonObject.has(key)) {
            return jsonObject.getJSONObject(key) ?: fallback
        }
        return fallback
    }

    override fun array(jsonObject: JSONObject, key: String, fallback: JSONArray?): JSONArray? {
        if (jsonObject.has(key)) {
            return jsonObject.getJSONArray(key) ?: fallback
        }
        return fallback
    }

    override fun string(jsonObject: JSONObject, key: String, fallback: String?): String? {
        if (jsonObject.has(key)) {
            return jsonObject.getString(key) ?: fallback
        }
        return fallback
    }

    override fun number(jsonObject: JSONObject, key: String, fallback: Number?): Number? {
        if (jsonObject.has(key)) {
            return jsonObject.get(key) as Number? ?: fallback
        }
        return null
    }

    override fun bool(jsonObject: JSONObject, key: String, fallback: Boolean?): Boolean? {
        if (jsonObject.has(key)) {
            return jsonObject.getBoolean(key)
        }
        return fallback
    }

    // from Array

    override fun get(jsonArray: JSONArray, index: Int, fallback: JSONObject?): JSONObject? {
        return try {
            jsonArray.getJSONObject(index) ?: fallback
        } catch (e: JSONException) {
            return fallback
        }
    }

    override fun array(jsonArray: JSONArray, index: Int, fallback: JSONArray?): JSONArray? {
        return try {
            jsonArray.getJSONArray(index) ?: fallback
        } catch (e: JSONException) {
            return fallback
        }
    }

    override fun string(jsonArray: JSONArray, index: Int, fallback: String?): String? {
        return try {
            jsonArray.getString(index) ?: fallback
        } catch (e: JSONException) {
            return fallback
        }
    }

    override fun number(jsonArray: JSONArray, index: Int, fallback: Number?): Number? {
        return try {
            jsonArray.get(index) as? Number? ?: fallback
        } catch (e: JSONException) {
            return fallback
        }
    }

    override fun bool(jsonArray: JSONArray, index: Int, fallback: Boolean?): Boolean? {
        return try {
            jsonArray.getBoolean(index)
        } catch (e: JSONException) {
            return fallback
        }
    }

}
